src/pydocstyle/parser.py

```diff
@@ -117,7 +118,29 @@ def is_public(self):
 
         This helps determine if it requires a docstring.
         """
-        return not self.name.startswith('_') or self.name.startswith('__')
+        module_name = Path(self.name).stem
+        return (
+            not self._is_inside_private_package() and
+            self._is_public_name(module_name)
+        )
+
+    def _is_inside_private_package(self):
+        """Return True if the module is inside a private package."""
+        path = Path(self.name)
+        package_names = path.parts[:-1]
```

Comment:
I think there's a fundamental issue here. The publicity of a module doesn't depend on its _entire_ path, just the path leading to directory inside one of the paths in `PYTHONPATH`.

So let's say I have a module in path `/_foo/bar/baz.py`. If `PYTHONPATH` is set to `/`, then `baz.py` is private. If `PYTHONPATH` is set to `/_foo/` then `baz.py` is public.

Fortunately `sys.path` is a list of all the filepaths in `PYTHONPATH`, so this can be relatively simply implemented. Something like:

```py
def _is_inside_private_package(name):
    path = Path(name).parent  # Ignore the actual module's name
    syspath = [Path(p) for p in sys.path]  # Convert to pathlib.Path.

    while path != path.parent and path not in syspath:  # Bail if we are at the root directory or in `PYTHONPATH`.
        if _is_private_name(path.name):
            print(f"Stopping at private package: {path}")
            return True
        path = path.parent

    return False
```

Reply:
Good catch! Let me try and incorporate your suggestion, and ideally add a unit test to capture this.

Reply:
Please let me know if https://github.com/PyCQA/pydocstyle/pull/494/commits/4b57b42da6007dba708ad0052189a9e0bebda426 addresses your comment, and if there are any other changes desired.